Create COMException from IRestrictedErrorInfo instead of calling Marshal.GetExceptionForHR.

jlaanstra opened the pull request:
Fixes #1334 

This change manually creates the COMException from the IRestrictedErrorInfo. It also updates the message for all exceptions to include the more useful restrictedError as .NET used to do when it had WinRT support.

This change currently doesn't handle any .NET specific HResults which Marshal.GetExceptionForHR does. We could potentially check for IErrorInfo if there is no IRestrictedErrorInfo and still call Marshal.GetExceptionForHR in that case as it's unlikely to be a WinRT call at that point. Thoughts?

manodasanW commented:
For the most part, this seems fine to me.  @AaronRobinsonMSFT feel free to comment if you see anything `Marshal.GetExceptionForHr` handled before that this change doesn't handle.

My only concern is lets say a DirectoryNotFoundException is thrown in C#.  This I assume gets translated to STG_E_PATHNOTFOUND based on [this](https://github.com/microsoft/CsWinRT/blob/04a590360338092bed7f80355e51a9af9950b022/src/WinRT.Runtime/ExceptionHelpers.cs#LL261C1-L262C1) and [this](https://learn.microsoft.com/en-us/dotnet/api/system.io.directorynotfoundexception?view=net-7.0#remarks).  If this exception was to cross the ABI boundary and then end up in C# code again in the same process that probably would be fine given we will retrieve the managed exception from the error info and rethrow it.  But if this was from a different process, we won't do that today and then when we do the error code translation, I think we are going to end up with a ComException.  I assume before we would have ended up with a DirectoryNotFoundException (needs to be confirmed).  Some of the other .NET errors are probably less common, but I feel like there might be some that are common like this one.

All other calls to [Marshal.GetExceptionForHr](https://github.com/search?q=repo%3Amicrosoft%2FCsWinRT+Marshal.throwexceptionforhr&type=code) probably needs to be updated to go through ExceptionHelpers.GetExceptionForHr, but this PR doesn't need to necessarily address that.

jlaanstra commented:
> For the most part, this seems fine to me. @AaronRobinsonMSFT feel free to comment if you see anything `Marshal.GetExceptionForHr` handled before that this change doesn't handle.
> 
> My only concern is lets say a DirectoryNotFoundException is thrown in C#. This I assume gets translated to STG_E_PATHNOTFOUND based on [this](https://github.com/microsoft/CsWinRT/blob/04a590360338092bed7f80355e51a9af9950b022/src/WinRT.Runtime/ExceptionHelpers.cs#LL261C1-L262C1) and [this](https://learn.microsoft.com/en-us/dotnet/api/system.io.directorynotfoundexception?view=net-7.0#remarks). If this exception was to cross the ABI boundary and then end up in C# code again in the same process that probably would be fine given we will retrieve the managed exception from the error info and rethrow it. But if this was from a different process, we won't do that today and then when we do the error code translation, I think we are going to end up with a ComException. I assume before we would have ended up with a DirectoryNotFoundException (needs to be confirmed). Some of the other .NET errors are probably less common, but I feel like there might be some that are common like this one.
> 
> All other calls to [Marshal.GetExceptionForHr](https://github.com/search?q=repo%3Amicrosoft%2FCsWinRT+Marshal.throwexceptionforhr&type=code) probably needs to be updated to go through ExceptionHelpers.GetExceptionForHr, but this PR doesn't need to necessarily address that.

What would the IRestrictedErrorInfo look like when the exception is coming from another process?

manodasanW commented:
> > For the most part, this seems fine to me. @AaronRobinsonMSFT feel free to comment if you see anything `Marshal.GetExceptionForHr` handled before that this change doesn't handle.
> > My only concern is lets say a DirectoryNotFoundException is thrown in C#. This I assume gets translated to STG_E_PATHNOTFOUND based on [this](https://github.com/microsoft/CsWinRT/blob/04a590360338092bed7f80355e51a9af9950b022/src/WinRT.Runtime/ExceptionHelpers.cs#LL261C1-L262C1) and [this](https://learn.microsoft.com/en-us/dotnet/api/system.io.directorynotfoundexception?view=net-7.0#remarks). If this exception was to cross the ABI boundary and then end up in C# code again in the same process that probably would be fine given we will retrieve the managed exception from the error info and rethrow it. But if this was from a different process, we won't do that today and then when we do the error code translation, I think we are going to end up with a ComException. I assume before we would have ended up with a DirectoryNotFoundException (needs to be confirmed). Some of the other .NET errors are probably less common, but I feel like there might be some that are common like this one.
> > All other calls to [Marshal.GetExceptionForHr](https://github.com/search?q=repo%3Amicrosoft%2FCsWinRT+Marshal.throwexceptionforhr&type=code) probably needs to be updated to go through ExceptionHelpers.GetExceptionForHr, but this PR doesn't need to necessarily address that.
> 
> What would the IRestrictedErrorInfo look like when the exception is coming from another process?

I assume it would be some form of the one created [here ](https://github.com/microsoft/CsWinRT/blob/04a590360338092bed7f80355e51a9af9950b022/src/WinRT.Runtime/ExceptionHelpers.cs#L236) with some error information possibly removed during marshaling based on access permissions of the process.  If none were removed, today, that would mean it would have the restricted error and a proxy to the managed exception.  The latter is on my list to look at if there is a better way to handle that to allow reusing the managed exception.  But in both cases, the HRESULT that was originally used for `RoOriginateLanguageException` would be in the `IRestrictedErrorInfo`.

jlaanstra commented:
Need to map at least the following set of COR_E hresults that are effectively aliases:

#define COR_E_UNAUTHORIZEDACCESS E_ACCESSDENIED
#define COR_E_ARGUMENT E_INVALIDARG
#define COR_E_INVALIDCAST E_NOINTERFACE
#define COR_E_OUTOFMEMORY E_OUTOFMEMORY
#define COR_E_NULLREFERENCE E_POINTER
#define COR_E_ARITHMETIC __HRESULT_FROM_WIN32(ERROR_ARITHMETIC_OVERFLOW)
#define COR_E_PATHTOOLONG __HRESULT_FROM_WIN32(ERROR_FILENAME_EXCED_RANGE)
#define COR_E_FILENOTFOUND __HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND)
#define COR_E_ENDOFSTREAM __HRESULT_FROM_WIN32(ERROR_HANDLE_EOF)
#define COR_E_DIRECTORYNOTFOUND __HRESULT_FROM_WIN32(ERROR_PATH_NOT_FOUND)
#define COR_E_STACKOVERFLOW __HRESULT_FROM_WIN32(ERROR_STACK_OVERFLOW)
#define COR_E_AMBIGUOUSMATCH _HRESULT_TYPEDEF_(0x8000211DL)
#define COR_E_TARGETPARAMCOUNT _HRESULT_TYPEDEF_(0x8002000EL)
#define COR_E_DIVIDEBYZERO _HRESULT_TYPEDEF_(0x80020012L)
#define COR_E_BADIMAGEFORMAT _HRESULT_TYPEDEF_(0x8007000BL)

jlaanstra commented:
Updated to handle additional hresults.

Left these out as I wasn't sure about the mappings:
#define COR_E_AMBIGUOUSMATCH HRESULT_TYPEDEF(0x8000211DL)
#define COR_E_TARGETPARAMCOUNT HRESULT_TYPEDEF(0x8002000EL)
#define COR_E_DIVIDEBYZERO HRESULT_TYPEDEF(0x80020012L)

I added mappings for E_BOUNDS and ERROR_CANCELED

manodasanW commented on src/WinRT.Runtime/ExceptionHelpers.cs:
```diff
@@ -152,54 +169,119 @@ private static Exception GetExceptionForHR(int hr, bool useGlobalErrorState, out
+                    {
+                        // For cross language WinRT exceptions, general information will be available in the description,
+                        // which is populated from IRestrictedErrorInfo::GetErrorDetails and more specific information will be available
+                        // in the resrictedError which also comes from IRestrictedErrorInfo::GetErrorDetails. If both are available, we
```
```suggestion
                        // in the restrictedError which also comes from IRestrictedErrorInfo::GetErrorDetails. If both are available, we

```

AaronRobinsonMSFT commented:
This looks good. Sorry for the late reply.
